### ISSUE TITLE
[Bug] `Type[Self]` type hints raise unreadable warnings and exceptions that are awful :grimacing: 

### ISSUE THREAD

**Issue by msvensson222** [beartype/beartype#389]
Hey! 

Have been using beartype a lot lately, love it! So far, havent had any issues, but this one I can't really figure out :( 

I have defined a integer wrapper, so I can keep track of different "types" of indices (all integers), and avoid bugs. It's implemented as such:
```python
# app.utils.py

from typing import Any, Self, Type


class MyCustomInt(int):
    def __new__(cls: Type[Self], value: int) -> Self:
        if not isinstance(value, int):
            raise ValueError(f"{cls.__name__} values must be integers")
        return super(MyCustomInt, cls).__new__(cls, value)

    def __repr__(self) -> str:
        return f"{self.__class__.__name__}({int.__repr__(self)})"

    def __str__(self) -> str:
        return int.__str__(self)

class HourIdx(MyCustomInt):
    pass
    
class MinuteIdx(MyCustomInt):
    pass
    
# ... etc
```
Hence, when typehinting using `HourIdx` for a method argument for example, I have to pass an instance of that class, and not just any int, which makes bugs in my relatively large project less likely. I'm using `beartype_this_package`, as such:

``` python
# app/__init__.py
from beartype.claw import beartype_this_package

beartype_this_package()
```

## Problem
However, I'm running into the following beartype warning when importing the class:
```
from app.utils import MyCustomInt
```
```
BeartypeClawDecorWarning: <protocol "app.utils.MyCustomInt"> in file "/<root-path>/my-project/app/utils.py" line 3:
    beartype.roar._BeartypeUtilObjectNameException: <staticmethod(<function MyCustomInt.__new__ at 0x115299da0>)> unnamed (i.e., declares neither "__name__" nor "__qualname__" dunder attributes).
  warn(message, cls)
```

I really don't understand the warning, and if I can somehow "fix" it. I've tried search a lot, but to no avail. Any idea why beartype is complaining, and if I can do anything about it?

**Versions:**
Beartype: 0.18.5   
Python: 3.11

Thanks in advance!

**Comment by leycec** [comment]
Thanks so much for the kind words! And... **you're absolutely right,** @msvensson222. That warning message truly is trash, is it not? Thankfully, you can make that immediately go away. Just do as @leycec says, not as @leycec does:

```python
class MyCustomInt(int):
    # Just comment this out problematic method signature out...
    #def __new__(cls: Type[Self], value: int) -> Self:

    # ...in favour of this awesome method signature that works!
    def __new__(cls, value: int) -> Self:
        # Also, note that you no longer need to explicitly validate that
        # "value" is an integer. @beartype will now do that for you. \o/
        #if not isinstance(value, int):
        #    raise ValueError(f"{cls.__name__} values must be integers")

        return super().__new__(cls, value)

    # Likewise, this is redundant. I think? Maybe? Probably. Let's pretend
    # @leycec knows what he's talking about here.
    #def __str__(self) -> str:
    #    return int.__str__(self)
```

Basically, it's `Type[Self]`. That's what I'm saying. That's the problematic type hint. When you decorate your `MyCustomInt` class manually by the `@beartype` decorator, `@beartype` helpfully explains this discrepancy with a readable exception:

```python
beartype.roar.BeartypeDecorHintPep484585Exception: Class method __main__.MyCustomInt.__new__()
parameter "cls" PEP 484 or 585 subclass type hint typing.Type[typing.Self] child type hint
typing.Self neither class, union of classes, nor forward reference to class.
```

Basically, `typing.Self` is a bit... *odd.* It doesn't necessarily work the way everyone would like. Thankfully, it's usually *never* necessary (or desirable) to annotate the first mandatory `cls` and `self` parameters of methods with type hints. Python already guarantees these parameters to have the correct types.

## Type Hints or GTFO, Bro

**Okay. Okay.** If you'd still like to annotate these sorts of things, you *absolutely* can – you just can't use `typing.Self` to do it. Instead, "just" use a forward reference-bounded type variable like so:

```python
# Oh – and prefer "beartype.typing" to "typing" to avoid deprecation warnings. *sigh*
from beartype.typing import Self, Type, TypeVar

TT = TypeVar('TT', bound='MyCustomInt')

class MyCustomInt(int):
    # This totally works too, if you'd prefer to go the Full Monty.
    def __new__(cls: Type[TT], value: int) -> Self:
        ...
```

Beartype: *now you're playing with confusing power.* :muscle: :bear:

**Comment by leycec** [comment]
**Wait.** I'm kinda dumb. `Type[Self]` should *totally* be fine. Therefore, so should your example – and, indeed, your codebase. Clearly, I botched that interaction. Let's escalate this to a full-blown issue. :facepalm:

**Comment by msvensson222** [comment]
Thanks for the extensive (and fast!) reply, appreciate it! As you point out, perhaps type-hinting `cls` is a bit overkill anyways, so I guess I can live without that. To be honest I didn't really get that the warning came from that specific type hint, and thought it had something to do with my implementation of the functionality of `MyCustomInt`. But know I know! Thanks again :)

**Comment by leycec** [comment]
> To be honest I didn't really get that the warning came from that specific type hint...

**Oh, totally.** Thankfully, none of this is your fault. In fact, you're amazing. You submitted a super-detailed issue publicly exhibiting our failures. I'm grateful for that.

That warning message is an unreadable mess – which is bad. Moreover, `Type[Self]` doesn't even appear to be working – which is worse than bad. *How did it come to this, @beartype!?* `</sigh>`

**Comment by leycec** [comment]
**Resolved by e3f322bb0df.** @beartype now explicitly supports `Type[Self]` type hints. Rejoice, everybody! @leycec did something. :smiling_face_with_tear:
